src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+            {
+                InsertText(textChanges, selectedSpan.Start, commentInfo.BlockCommentEndString);
+                InsertText(textChanges, selectedSpan.Start, commentInfo.BlockCommentStartString);
+                // Shrink the tracking so the previous comment start marker is not included in selection.
+                amountToAddToStart = commentInfo.BlockCommentEndString.Length;
+            }
+
+            // If the end is commented (and not a comment marker), close the current comment and open a new one.
+            if (blockCommentSelection.IsLocationCommented(selectedSpan.End)
+                && !blockCommentSelection.DoesEndWithBlockComment(commentInfo))
+            {
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentEndString);
+                InsertText(textChanges, selectedSpan.End, commentInfo.BlockCommentStartString);
+                // Shrink the tracking span so the next comment start marker is not included in selection.
+                amountToAddToEnd = -commentInfo.BlockCommentStartString.Length;
+            }
+
+            trackingSpans.Add(new CommentTrackingSpan(selectedSpan, amountToAddToStart, amountToAddToEnd));
+        }
+
+        private static void AddBlockComment(CommentSelectionInfo commentInfo, TextSpan span, List<TextChange> textChanges)
+        {
+            InsertText(textChanges, span.Start, commentInfo.BlockCommentStartString);
+            InsertText(textChanges, span.End, commentInfo.BlockCommentEndString);
+        }
+
+        private static void DeleteBlockComment(BlockCommentSelectionHelper blockCommentSelection, TextSpan spanToRemove,
+            List<TextChange> textChanges, CommentSelectionInfo commentInfo)
+        {
+            DeleteText(textChanges, new TextSpan(spanToRemove.Start, commentInfo.BlockCommentStartString.Length));
+            var endMarkerPosition = spanToRemove.End - commentInfo.BlockCommentEndString.Length;
+            // Sometimes the block comment will be missing a close marker.
+            if (Equals(blockCommentSelection.GetSubstringFromText(endMarkerPosition, commentInfo.BlockCommentEndString.Length),
+                commentInfo.BlockCommentEndString))
+            {
+                DeleteText(textChanges, new TextSpan(endMarkerPosition, commentInfo.BlockCommentEndString.Length));
+            }
+        }
+
+        private class BlockCommentSelectionHelper
+        {
+            private readonly string _text;
+            private readonly ITextSnapshot _snapshot;
+
+            public TextSpan SelectedSpan { get; }
+
+            public IEnumerable<TextSpan> IntersectingBlockComments { get; }
+
+            public IEnumerable<TextSpan> UncommentedSpansInSelection { get; }
+
+            public BlockCommentSelectionHelper(IEnumerable<TextSpan> allBlockComments, SnapshotSpan selectedSnapshotSpan)
+            {
+                _text = selectedSnapshotSpan.GetText().Trim();
```

Comment:
why trim?

Reply:
text here is mainly used to check the beginning and end of selection for comment markers.  In those cases we don't care about whitespace.  I'll rename the variable as such.

Reply:
yeah, this coudl def use some clarity.  esp. as you end up examining this trimmed text, but then using the original span for a lot of other things.  it would be good if it was clear that we're not mixing/matching between both.